File: python/exercise44/product_search.py
import json

products: list[dict[str, str | float]] = []


def search_product(name: str):
    for product in products:
        if str(product["name"]).lower() == name.lower():
            return product

    return None


def display_product(product: dict[str, str | float]):
    name, price, quantity = product.values()

    print(f"Name: {name}")
    print(f"Price: ${float(price):.2f}")
    print(f"Quantity on hand: {int(quantity)}")


def add_product(name: str):
    price = input("Product price: ").strip()
    quantity = input("Product quantity: ").strip()

    products.append({"name": name, "price": float(price), "quantity": int(quantity)})


with open("./products.json") as json_file:
    contents = json_file.read()
    products = json.loads(contents)["products"]


while True:
    product_name = input("What is the product name? ").strip()

    found = search_product(product_name)
    if not found:
        print("Sorry, that product was not found in out inventory.")
        response = input("Want to add the product? (y/n)? ")

        if response.strip() == "y":
            add_product(product_name)

        print()
        continue

    display_product(found)
    break

File: javascript/exercise44/product-search.js
import fs from "node:fs";
import { input } from "pynput";

const contents = fs.readFileSync("./products.json", "utf8");

/** @type {{name: string, price: number, quantity: number}[]} */
const products = JSON.parse(contents).products;

while (true) {
	const product_name = await input("What is the product name? ");
	const found = find_product(product_name);

	if (found) {
		display_product(found);
		break;
	}
	console.log("Sorry, that product was not found in our inventory.");

	const valid = await input("Add the product? (y/n)? ");
	if (valid.trim() === "y") {
		await add_product(product_name);
	}
}

/** @param {{name: string, price: number, quantity: number}} product  */
function display_product(product) {
	console.log(`Name: ${product.name}`);
	console.log(`Price: $${product.price.toFixed(2)}`);
	console.log(`Quantity on hand: ${product.quantity}`);
}

/** @param {string} name  */
function find_product(name) {
	return products.find(
		(product) => product.name.toLowerCase() === name.toLowerCase(),
	);
}

/** @param {string} name */
async function add_product(name) {
	const price = await input("Price: ", { convert: "float" });
	const quantity = await input("Quantity: ", { convert: "int" });
	products.push({ name, price, quantity });
	console.log();
}
